Added monitored directories to project panel (Change request #2024)

Arkarit opened the pull request:
Monitored directories can be added to a project.
Each monitored directory can have multiple search patterns. These search patterns are displayed in the directory and in the info tip. Clearing the search pattern filters for all files. Sub directories, which contain no data matching the pattern, are hidden.
Directories can also be renamed manually.


Implementation details:
The monitoring is done by polling every second.

I don't really like polling, but I tried two other obvious attempts, both of which had severe drawbacks:

FindFirstChangeNotification/FindNextChangeNotification: Did not work very good, because a max of ~60 directories per thread can be handled (64 objects limit of WaitForMultipleObjects). Additionally, it is not very useful when used for full sub trees, because it does not offer details about WHAT was changed.

GetQueuedCompletionStatus/ReadDirectoryChangesW: Worked absolutely perfect in terms of being notified, BUT locked all open directories against deletion, which is definitely not what we want: The feature is mainly thought for directories, which are externally changed, e.g. a Perforce directory of LUA files, which change on every get latest version. It would be completely counterproductive if Perforce would be hindered from deleting directories, because Notepad++ watches it.

The polling itself is quick, because the last write time is checked first.

Every 10 seconds, the visibility of the sub directories is checked.
Since this is a recursive operation (a sub-directory may not be hidden, if a sub-sub-sub-directory contains data), this is a far more time consuming operation. If you add directories with an extreme amount of sub-directories (e.g. C:\Windows\winsxs, it has ~18.000 sub directories on my machine!), a remarkably time is spent by checking these. Nevertheless, monitoring winsxs is not what the project panel is thought for, and for moderately big directories, it works pretty good.

Additionally, there is still some optimization potential for opening directories, which is slow currently for really large directories. Currently, if a directory changes, the directory is read and compared in the watcher thread, then a message is sent, and then its read and compared again in the project panel. It would be better, if the directory contents would be directly sent.

Probably, the monitoring feature and the sub directory visibility feature could be switched by user options. This is prepared, but not tested.

Feedback appreciated.

dail8859 commented:
+1 for the feature. I don't care for the polling idea either but as you said it can be pretty quick. You may want to take a look at some of the source code for TortoiseGit. It has a process ([TGitCache](GetQueuedCompletionStatus/ReadDirectoryChangesW)) that monitors directories for changes. Personally I have no idea how it works or if it would be useable for you at all, but grepping through the source code I did find references to GetQueuedCompletionStatus and ReadDirectoryChangesW.

Arkarit commented:
Ah... thanks for the hint with TGitCache.

They do it that way, that they only watch the base paths of the git directories recursively, and all other sub directories below the base directory are eveluated manually by the ReadDirectoryChangesW messages.

This could also be a good solution for Notepad++... only the directories manually added by the user would be locked, but everything below (the opened nodes) would be freely changeable. Sounds like a better solution than polling, I will check it on the weekend. Nevertheless, this comment I found in TGit makes me frown a bit:

// NOTE: the longer this code takes to execute until ReadDirectoryChangesW
// is called again, the higher the chance that we miss some
// changes in the file system!

I don't want to lose any change.

dail8859 commented:
I pulled this code and built it. Seemed to work fine from what I could tell -- not that I put it through anything rigorous. 

My one comment about it is that it is a bit confusing between adding a "folder" or adding a "directory". They sound the same but behave differently and even just testing it I kept getting them mixed up alot. Unfortunately I can't think of any good suggestions at this point...if I think of anything, I'll let you know.

ariccio commented on PowerEditor/src/WinControls/FunctionList/functionListPanel.cpp:
```diff
@@ -281,7 +281,9 @@ void FunctionListPanel::reload()
 	if (root)
 	{
-		_treeView.setItemParam(root, fullFilePath);
+		FunctionListPanelData* data = (FunctionListPanelData*)_treeView.getData(root);
```
Please, no C-style casts in new code.

ariccio commented on PowerEditor/src/WinControls/FunctionList/functionListPanel.h:
```diff
@@ -91,6 +91,20 @@ struct TreeParams {
+	FunctionListPanelData(const TCHAR* str) : TreeViewData(), _str(str) {}
+	virtual ~FunctionListPanelData() {}
+
+	virtual TreeViewData* clone() const {
```
Mark it `override`?

ariccio commented on PowerEditor/src/WinControls/FunctionList/functionListPanel.h:
```diff
@@ -91,6 +91,20 @@ struct TreeParams {
+	generic_string _str;
+
+	FunctionListPanelData(const TCHAR* str) : TreeViewData(), _str(str) {}
+	virtual ~FunctionListPanelData() {}
```
Maybe we can use the `= default` destructor?

ariccio commented on PowerEditor/src/WinControls/FunctionList/functionListPanel.h:
```diff
@@ -91,6 +91,20 @@ struct TreeParams {
+public:
+	generic_string _str;
+
+	FunctionListPanelData(const TCHAR* str) : TreeViewData(), _str(str) {}
```
How about a  `PCTSTR str` instead?

ariccio commented on PowerEditor/src/WinControls/FunctionList/functionListPanel.h:
```diff
@@ -91,6 +91,20 @@ struct TreeParams {
 	SearchParameters _searchParameters;
 };
 
+class FunctionListPanelData : public TreeViewData {
```
Is anybody going to derive from this class? Mark it `final`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,450 @@
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	_filters.push_back(generic_string(TEXT("*.*")));
```
How about:

```
_filters.emplace_back(TEXT("*.*"));
```

...instead?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,450 @@
+		append(_path, TEXT("*.*"), false);
+	else
+		for (size_t i=0; i<_filters.size(); ++i)
+			append(_path, _filters[i], false);
```
This is triply-nested. Braces, please?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,450 @@
+
+void Directory::append(const generic_string& path, const generic_string& filter, bool readDirs)
+{
+	bool readFiles = !readDirs;
```
This is a constant, I assume?

[Mark it `const`!](http://devopsreactions.tumblr.com/post/104400337816/mark-it-const)

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,450 @@
+
+	generic_string searchPath = _path + TEXT("\\") + filter;
+
+	WIN32_FIND_DATAW fd;
```
Why `WIN32_FIND_DATAW`, not `WIN32_FIND_DATA`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,450 @@
+
+	WIN32_FIND_DATAW fd;
+
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, FIND_FIRST_EX_LARGE_FETCH);
```
`FIND_FIRST_EX_LARGE_FETCH` is *not a performance win*, and I've actually [measured it to be *slower*](http://ariccio.com/2015/01/13/how-fast-are-findfirstfilefindfirstfileex-and-cfilefind-actually/).

ariccio replied:
*imagine that this comment is on the line*:

```
if (fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
```
...*because my phone is misbehaving*.

Make this explicit? i.e.:

```
if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
```

Arkarit commented:
Thank you for the suggestions :)
I have implemented them.
In ProjectPanel.cpp, I changed all LPARAM casts to be consistent there.

@dail8859:
Perhaps we should call it "Monitored Directories" just to clarify the difference to "Folders", which is indeed confusing now.

ariccio commented:
> Thank you for the suggestions :)

Your welcome! Always happy to help 😊

Sadly, I've been away from home, thus using my phone, and had to stop *right* in the middle of the code review, because my phone decided that it didn't want to leave any more line notes >:(

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+		const generic_string file(fd.cFileName);
```
Constructing a `std::wstring` can be quite expensive (it requires a heap allocation for strings more than 8 characters long). Might as well delay construction until actually needed. I will elaborate in further comments.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			if (file == TEXT(".") || file == TEXT(".."))
```
Here, you could do:

```
if( ( _tcscmp( fd.cFileName, TEXT("..") ) == 0 ) || ( _tcscmp( fd.cFileName, TEXT(".") ) == 0 ) )
```
...so that in the case that we're in an empty directory, we break out of the loop with only two (stack-allocated) string comparisons. If it's a non-empty directory, this is a faster path for the first result returned.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+					{
+						_invisibleDirs.push_back(file);
+						if (!FindNextFile(hFind, &fd))
+							break;
```
How many levels of indentation are we at here? This control structure is giving me a serious headache.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+		_filters = filters;
+
+	// first, read directories, never filtered.
+	append(_path, TEXT("*.*"), true);
```
Is `Directory::append` ever called with a **non**-member `path`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+		read(path);
+}
+
+bool Directory::read(const generic_string& path, const std::vector<generic_string>& filters)
```
Is `Directory::read(const generic_string& path, const std::vector<generic_string>& filters)` ever called with a **non**-member `path`?

ariccio replied:
Is `Directory::read(const generic_string& path, const std::vector<generic_string>& filters)` ever called with a **non**-member `filters`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+	_wasRead = true;
+
+	return _exists;
```
How about, instead of setting a member-variable `_exists`, then calling `append` (which modifies the value, *but* also returns `void`), simply returning a bool from `append`?

Arkarit replied:
I think because append() is called three times here and the return value would have to be or'ed each time, it's better to set the member variable. Additionally, it is a quick way to compare existing and non-existing directories, and can be accessed via getter.

ariccio replied:
> the return value would have to be or'ed each time

Huh? You can't do something like:

```

const bool path_appended = append( /* function args */ );
if (!path_appended) 
{
    readLastWriteTime(_lastWriteTime);
    return false;
}
readLastWriteTime(_lastWriteTime);
return true;

```  
...?


> Additionally, it is a quick way to compare existing and non-existing directories, and can be accessed via getter.

Ok, that I understand.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+// read last write time reads the last write time into "filetime".
+// Under some circumstances, it might happen that the file time can not be determined, in this case it returns false.
+bool Directory::readLastWriteTime(FILETIME& filetime) const
```
Is `filetime` intended to be an `_Out_` parameter? (yes, I think it is?) Mark it `_Out_`.

It's probably also a good idea (when marking it `_Out_`), to mark `Directory::readLastWriteTime` as such:

```
_Success_( return == true )
bool Directory::readLastWriteTime(_Out_ FILETIME& filetime) const
```
...so that static analysis can pick up misuse of the result (on a failed function call).

ariccio replied:
----
Personal preference/nitpick alert!

I happen to like the [Google style of *no non-const references*](https://youtu.be/NOCElcMcFik?t=1707) for `_Out_` parameters. Feel free to ignore me here.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+					generic_string childPath = path + TEXT("\\") + file;
+					if (!containsData(childPath))
+					{
+						_invisibleDirs.push_back(file);
```
If you use `emplace_back`, then you can do:

```
_invisibleDirs.emplace_back(fd.cFileName);
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
```
As a side effect of the complex control structures, I'm not sure if execution will ever reach the `if` statement where `hFind != INVALID_HANDLE_VALUE`, because that's the condition of the loop itself! So....?

ariccio replied:
Also a nitpick:

Maybe you ought to check the return value of `FindClose`, and `throw std::runtime_error` (or perhaps a `throw std::logic_error`) on failure?

Arkarit replied:
Thanks, the check for `INVALID_HANDLE_VALUE` does indeed not make any sense, neither in the loop condition nor here, since it was already checked before the loop entrance. Will make this whole loop a `do while`.

I don't think throwing is a good idea, since the only realistic reason, that `FindClose()` can fail, imho is a faulty handle, which would be caused by a previous error. So I think an `assert(0)` would be better here.

ariccio replied:
A `FindClose` failure in release builds is still a bug. If we ever decide to enable `ProcessStrictHandleCheckPolicy` (I hope we do!), then handle usage problems will (rightfully) trigger a `__fastfail`.

It's really better to detect things like this - an API that's documented to return a failure code, actually returning a failure code - as soon as possible, so the bug doesn't get swept under the rug!

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+bool Directory::containsDataChanged() const
+{
+	for (auto it=_dirs.begin(); it != _dirs.end(); ++it)
```
Any reason why you've used `C++03` style iteration here? Not `C++11` style:

```
for ( auto& dir : _dirs )
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+	if (_filters.empty())
+		return containsData(path, TEXT("*.*"));
+
+	for (size_t i=0; i<_filters.size(); ++i)
```
Nitpick:

That's a function call in loop bound!

why not:

```
const size_t filtersSize = _filters.size();
for (size_t i = 0; i < filtersSize; ++i)
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+bool Directory::containsData(const generic_string& path, const generic_string& filter) const
+{
+	generic_string searchPath = path + TEXT("\\") + filter;
+	generic_string childPath;
```
This isn't `C89`, there's no need to declare variables at the beginning of scope! */grumpy*

Declare it on line 270, where it's initialized. I don't expect that to be controversial, as it makes the code easier to read.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+{
+	generic_string searchPath = path + TEXT("\\") + filter;
+	generic_string childPath;
+	generic_string file;
```
Same here.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+	if (!readLastWriteTime(newFiletime))
+		return true;
+
+	// we could successfully get the filetime, but before this it was 0 - definitely a change
```
If you'd like to make this assumption, then maybe you should initialize `newFiletime` (at 297) like this:

```
FILETIME newFiletime = { 0 };
```
...as it won't cost you anything, and ensures that even if someone breaks something in `readLastWriteTime`, this assumption remains true;

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+	// compare file times
+	return _lastWriteTime.dwLowDateTime != newFiletime.dwLowDateTime
+		|| _lastWriteTime.dwHighDateTime != newFiletime.dwHighDateTime;
```
Nitpick: 

```
return   ( _lastWriteTime.dwLowDateTime   != newFiletime.dwLowDateTime )
	  || (  _lastWriteTime.dwHighDateTime != newFiletime.dwHighDateTime );
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+
+void Directory::setFilters(const std::vector<generic_string>& filters, bool autoread)
+{
+	if (_filters != filters)
```
So, you're checking each object in the entire vector for equality, and then assigning if they're different? Hmm. That might be expensive. It's also not clear if there's any benefit.

Arkarit replied:
It has a clear benefit only when `autoread` is true, because there is also an expensive `read()` in this case. Will change it so, that the check is only performed on `autoread`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -784,6 +1289,64 @@ HTREEITEM ProjectPanel::addFolder(HTREEITEM hTreeItem, const TCHAR *folderName)
 	return addedItem;
 }
 
+HTREEITEM ProjectPanel::addDirectory(HTREEITEM hTreeItem, const TCHAR *folderName, bool root, const TCHAR *monitorPath, const std::vector<generic_string>* filters)
```
So, `filters` is an optional parameter? Probably a good idea to mark it `_In_opt_`.

ariccio replied:
The `const TCHAR*` arguments should be one of:

- `_In_z_ const TCHAR*`
- `PCTSTR`

Arkarit replied:
I am not sure... agreed, that this change would be good, but `TCHAR*` `const TCHAR*` is used all over the source. 
Imho it probably should be better consistent currently, and refactored in a separate commit for all occurrences?

ariccio replied:
> Imho it probably should be better consistent currently, and refactored in a separate commit for all occurrences?

Normally, yes..... But this is *new* code and the `Notepad++` devs are, at times, quite hostile to code cleanup/refactoring.

milipili replied:
Ok that's enough @ariccio. I can not let you say this kind of things. We are not hostile to code cleanup and *certainly* not to refactoring since it is the current direction. It is because of the current refactoring that we can not accept all PR. I know, it takes time, but we do what we can. I think that you would already have understood that. You disagree with our methods and I understand that and maybe we're wrong. However, I don't know what you're looking for, but please calm down a bit...

ariccio replied:
I'm sorry for misrepresenting your position. I'm actually quite glad that I'm wrong here. This is very different than the "no refactoring/cleanup PRs will be accepted!"  that I've previously encountered.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -600,34 +1084,41 @@ void ProjectPanel::notified(LPNMHDR notification)
 				{
-					str = (generic_string *)lpGetInfoTip->lParam;
-					if (!str)
+					_infotipStr = ((ProjectPanelData *)lpGetInfoTip->lParam)->_filePath;
```
Oops, you missed a C-style cast here.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+	bool writeTimeHasChanged() const;
+
+	void setFilters(const std::vector<generic_string>& filters = std::vector<generic_string>(), bool autoread=true);
+	const std::vector<generic_string>& getFilters() const { return _filters; }
```
How many filters are we expecting to be in `_filters`? If it's small, it might be better to just make a copy, to avoid reference lifetime issues.

Arkarit replied:
It's not many filters, and they are small. But it is currently accessed often in `DirectoryWatcher`, so I tend to keep it as it is.

ariccio replied:
Okay, just be careful :)

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+		{
+			if (!FindNextFile(hFind, &fd))
+				return false;
+			continue;
```
Is the `continue` necessary? There's no other path that the loop could take?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+		}
+		else
+		{
+			// data was found - return true. This should never happen, because the files were already checked. Just to be safe.
```
> This should never happen, because the files were already checked. Just to be safe.

Seems like a good place to do an `assert( false );`?

Arkarit replied:
Oops... no. The assumption that this can not happen is wrong, because a file could be created between the 1st and 2nd iteration. I'll change the comment.

ariccio replied:
> The assumption that this can not happen is wrong, because a file could be created between the 1st and 2nd iteration. 

Oh, yay! Race conditions! So much fun!

A lot of the code in this function seems like duplicate code, which contributes to confusing problems like that sort of race condition. Surely there's a more elegant way to do it?

ariccio replied:
> The assumption that this can not happen is wrong, because a file could be created between the 1st and 2nd iteration. 

Oh, yay! Race conditions! So much fun!

A lot of the code in this function seems like duplicate code, which contributes to confusing problems like that sort of race condition. Surely there's a more elegant way to do it?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+			onFileRemoved(*it);
+	for (auto it = other._files.begin(); it != other._files.end(); ++it)
+		if (_files.find(*it) == _files.end())
+			onFileAdded(*it);
```
What's up with the `C++03` style loops?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+	if (PathIsRoot(_path.c_str()))
+	{
+		HANDLE hDir = CreateFile(_path.c_str(), FILE_LIST_DIRECTORY, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+		if (hDir != INVALID_HANDLE_VALUE)
```
I think this'd be clearer if you swapped:

```
if (hDir != INVALID_HANDLE_VALUE)
{
    //If hDir is valid
    //...
    return true;
}
return false;
```
for:
```
if (hDir == INVALID_HANDLE_VALUE)
{
    //If hDir is invalid
    return false;
}
//If hDir is valid
//...
return true;
```
Does that make sense?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.cpp:
```diff
@@ -0,0 +1,454 @@
+	bool result = false;
+	HANDLE hToken;
+
+	if (OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken))
```
Also might be clearer here if you returned early like:

```
const BOOL openResult = OpenProcessToken(::GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken);
if ( openResult == 0 )
{
    return false;
}
//Successful path
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+	// directory with path - is instantly read in ctor, if autoread==true
+	Directory(const generic_string& path, const std::vector<generic_string>& filters = std::vector<generic_string>(), bool hideEmptyDirs=true, bool autoread=true );
+
+	virtual ~Directory() {}
```
`= default;`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+
+	// empty directories can be hidden:
+	// "empty" means in this case, neither the directory itself nor its subdirectories contains any data, which match the filters
+	bool getHideEmptyDirs() const { return _hideEmptyDirs; }
```
`getHideEmptyDirs` is not used? I also don't see the purpose of a one-line-getter for a `bool`. That doesn't improve encapsulation?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+	virtual void onDirRemoved(const generic_string&) {}
+	virtual void onFileAdded(const generic_string&) {}
+	virtual void onFileRemoved(const generic_string&) {}
+	virtual void onEndSynchronize(const Directory&) {}
```
Are these supposed to be pure virtual? Why no `= 0;`?

Arkarit replied:
No, these are not pure virtual, and the base class Directory is used in the code without being inherited.

ariccio replied:
Ah, ok. A bit counterintuitive.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.h:
```diff
@@ -0,0 +1,154 @@
+	DWORD _checkEmptyDirsCount;
+
+
+	std::set<Directory*> _watchdirs;
```
Why the manual memory management? Why not a `std::set<std::unique_ptr<Directory>>`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.h:
```diff
@@ -0,0 +1,154 @@
+
+	std::set<Directory*> _watchdirs;
+	std::map<HTREEITEM, Directory*> _dirItems;
+	std::map<Directory*, int> _dirItemReferenceCount;
```
Same question for `_dirItems` and `_dirItemReferenceCount`? Maybe a `std::shared_ptr` would be better?

Arkarit replied:
I have to confess that I am not really used to smart pointers.
But it seems to be a very good idea, because it also removes the necessity to manually keep track of the reference count, since it comes for free with `shared_ptr`.
Will change it this way.

ariccio replied:
Yeah, smart pointers are the way to go.

Just be careful if you're storing any non-trivial/non-`pod` object in a `std::shared_ptr`, as you'll need to use [`std::enable_shared_from_this`](http://en.cppreference.com/w/cpp/memory/enable_shared_from_this), else you'll have all sorts of horrible memory corruption issues.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -36,17 +36,26 @@
+#define INDEX_OFFLINE_BASEDIR	11
+#define INDEX_LEAF_DIRFILE		12
 
 #define GET_X_LPARAM(lp)                        ((int)(short)LOWORD(lp))
```
[Wait, why are these defined manually?](https://msdn.microsoft.com/en-us/library/windows/desktop/ms632654.aspx)

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+};
+
+
+#endif
```
Grr.
![directory h_endl](https://cloud.githubusercontent.com/assets/2142308/8297005/dff55fbe-1925-11e5-8ece-1d1de4348a3e.png)

A new line at the end of the file is always a good idea.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -306,6 +376,42 @@ BOOL ProjectPanel::setImageList(int root_clean_id, int root_dirty_id, int projec
+	if (hbmp == NULL)
+		return FALSE;
+	ImageList_AddMasked(_hImaLst, hbmp, maskColour);
+	DeleteObject(hbmp);
```
Both `ImageList_AddMasked` and `DeleteObject` can fail. It's probably a good idea to check the return value, and `throw std::logic_error` on failure?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
+				tvItem.mask = TVIF_PARAM;
+				tvItem.hItem = hNewDirectory;
+				::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hParentItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -511,18 +719,19 @@ void ProjectPanel::openSelectFile()
 	NodeType nType = getNodeType(tvItem.hItem);
-	generic_string *fn = (generic_string *)tvItem.lParam;
-	if (nType == nodeType_file && fn)
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+		hMenu = NULL; // currently no context menu for monitored files
+	else //nodeType_file
+		hMenu = _hFileMenu;
+	return hMenu;
```
Why assign to a local variable? Why not just return the value itself? i.e.:

```
	NodeType nodeType = getNodeType(hTreeItem);
	if (nodeType == nodeType_root)
		return _hWorkSpaceMenu;
	else if (nodeType == nodeType_project)
		return _hProjectMenu;
	else if (nodeType == nodeType_folder)
		return _hFolderMenu;
	else if (nodeType == nodeType_baseDir)
		return _hDirectoryMenu;
	else if (nodeType == nodeType_dir || nodeType == nodeType_dirFile)
		return NULL; // currently no context menu for monitored files
	return _hFileMenu;
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+
+}
+
+void ProjectPanel::onTreeItemAdded(bool afterClone, HTREEITEM hItem, TreeViewData* newData)
```
`newData` is *not* an optional parameter? Maybe you should mark it as `_In_`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+				tvItem.hItem = hItem;
+				tvItem.mask = TVIF_PARAM;
+				SendMessage(hwnd, TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				treeItemChanged(hItem, (TreeViewData*) tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hTreeItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	{
+		case DIRECTORYWATCHER_UPDATE:
+		{
+			HTREEITEM hItem = (HTREEITEM)lParam;
```
Oops, you missed a C-style cast!

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	tvItem.cchTextMax = MAX_PATH;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+		if (data._userLabel.empty())
+		{
+			generic_string newFolderLabel = buildDirectoryName(data._filePath, data._filters);
+			wcsncpy(tvItem.pszText,newFolderLabel.c_str(), MAX_PATH-1);
```
- Why use `wcsncpy` when passing a `TCHAR[]`? (instead of `_tcsncpy`)
- You should use `_tcsncpy_s`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItemNode;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -590,9 +1054,29 @@ void ProjectPanel::notified(LPNMHDR notification)
+					{
+						data._userLabel.clear();
+						generic_string newFolderLabel = buildDirectoryName(data._filePath, data._filters);
+						wcsncpy(tvItem.pszText,newFolderLabel.c_str(), MAX_PATH-1);
```
- Why use `wcsncpy` when passing a `TCHAR[]`? (instead of `_tcsncpy`)
- You should use `_tcsncpy_s`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -600,34 +1084,41 @@ void ProjectPanel::notified(LPNMHDR notification)
 				}
-				lpGetInfoTip->pszText = (LPTSTR)str->c_str();
-				lpGetInfoTip->cchTextMax = str->size();
+				lpGetInfoTip->pszText = (LPTSTR)_infotipStr.c_str();
```
Oops, you missed a C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -699,33 +1215,14 @@ void ProjectPanel::setWorkSpaceDirty(bool isDirty)
 }
 
-NodeType ProjectPanel::getNodeType(HTREEITEM hItem)
+ProjectPanel::NodeType ProjectPanel::getNodeType(HTREEITEM hItem) const
```
Yay! It's `const` now!

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -699,33 +1215,14 @@ void ProjectPanel::setWorkSpaceDirty(bool isDirty)
-	}
+	tvItem.mask = TVIF_PARAM;
+	SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -772,9 +1261,25 @@ POINT ProjectPanel::getMenuDisplyPoint(int iButton)
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -784,6 +1289,64 @@ HTREEITEM ProjectPanel::addFolder(HTREEITEM hTreeItem, const TCHAR *folderName)
+	if (_selDirOfFilesFromDirDlg == TEXT("") && _workSpaceFilePath != TEXT(""))
+	{
+		TCHAR dir[MAX_PATH];
+		lstrcpy(dir, _workSpaceFilePath.c_str());
```
***never, ever, EVER use `lstrcpy`.***

["lstrcpy is a similarly named Microsoft design defect that appears to behave like strlcpy but is actually a security bug."](https://randomascii.wordpress.com/2013/04/03/stop-using-strncpy-already/)

["Copies a string to a buffer. **Warning**: Do not use."](https://msdn.microsoft.com/en-us/library/windows/desktop/ms647490.aspx)

Please replace with `_tcscpy_s( dir, workSpaceFilePath.c_str());`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+				{
+					_filters.clear();
+					const int filterSize = 256;
+					TCHAR tfilters[filterSize + 1];
```
You should zero-initialize `tfilters`:
```
TCHAR tfilters[filterSize +1] = { 0 };
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+					const int filterSize = 256;
+					TCHAR tfilters[filterSize + 1];
+					tfilters[filterSize] = '\0';
+					::GetDlgItemText(_hSelf, IDD_PROJECTPANEL_FILTERS_COMBO, tfilters, filterSize);
```
You ***absolutely*** should check to see if `::GetDlgItemText` returned `0` before using `tfilters`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+					TCHAR tfilters[filterSize + 1];
+					tfilters[filterSize] = '\0';
+					::GetDlgItemText(_hSelf, IDD_PROJECTPANEL_FILTERS_COMBO, tfilters, filterSize);
+					addText2Combo(tfilters, ::GetDlgItem(_hSelf, IDD_PROJECTPANEL_FILTERS_COMBO));
```
`::GetDlgItem` **can** fail.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+void FilterDlg::addText2Combo(const TCHAR * txt2add, HWND hCombo)
+{
+	if (!hCombo) return;
+	if (!lstrcmp(txt2add, TEXT(""))) return;
```
Please use `_tcscmp` instead.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+generic_string FilterDlg::getTextFromCombo(HWND hCombo, bool) const
+{
+	TCHAR str[PROJECTPANEL_FILTERS_MAXLENGTH];
+	::SendMessage(hCombo, WM_GETTEXT, PROJECTPANEL_FILTERS_MAXLENGTH - 1, reinterpret_cast<LPARAM>(str));
```
You should check to see if `::SendMessage` returned `0` before using `str`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+
+generic_string FilterDlg::getTextFromCombo(HWND hCombo, bool) const
+{
+	TCHAR str[PROJECTPANEL_FILTERS_MAXLENGTH];
```
You should zero-initialize `str`:
```
TCHAR str[PROJECTPANEL_FILTERS_MAXLENGTH] = { 0 };
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+	for (int i = 0; i < count; ++i)
+	{
+		::SendMessage(hCombo, CB_GETLBTEXT, i, reinterpret_cast<LPARAM>(text));
+		strings.emplace_back(generic_string(text));
```
You ***absolutely*** should check to see if `::SendMessage` returned `CB_ERR` before using `text`.

ariccio replied:
From the [documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/bb775862.aspx):

> Security Warning:  Using this message incorrectly can compromise the security of your program. This message does not provide a way for you to know the size of the buffer. If you use this message, first call CB_GETLBTEXTLEN to get the number of characters that are required and then call the message to retrieve the string. You should review the Security Considerations: Microsoft Windows Controls before continuing.

You should use `CB_GETLBTEXTLEN` ***instead of*** `CB_GETCOUNT`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+
+int FilterDlg::saveComboHistory(int id, int maxcount, std::vector<generic_string> & strings)
+{
+	TCHAR text[PROJECTPANEL_FILTERS_MAXLENGTH];
```
You should zero-initialize `text`:
```
TCHAR text[PROJECTPANEL_FILTERS_MAXLENGTH] = { 0 };
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+
+	tvItem.hItem = hItem;
+	SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+	, _treeView(&projectPanel->getTreeView())
+	, _hItem(hItem)
+{
+	TCHAR textBuffer[MAX_PATH];
```
You should zero-initialize `textBuffer`:
```
TCHAR textBuffer[MAX_PATH] = { 0 };
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+	{
+		tvItem.hItem = hItemNode;
+		SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+		ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+		return dataL.isDirectory() ? -1 : 1;
+
+	// both items are of the same kind.
+	return lstrcmpi(dataL._name.c_str(),dataR._name.c_str());
```
Please use `_tcsicmp`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+
+void ProjectPanel::TreeUpdaterDirectory::onEndSynchronize(const Directory& other)
+{
+	other;
```
What the hell?

Arkarit replied:
This is just a way to overcome
"c4100 unreferenced formal parameter".
I prefer this way, since it clearly shows that this parameter is not used, without removing its meaning by removing the name from the parameter list.

Anyway, I do not see the usefulness of this warning at all. It makes things far less readable without any real benefit (especially in optional virtual functions). I suggest to switch it off globally.

But of course this is not my project ;), I can change it to removing the name(s) from the parameter list(s), if you like this better.

ariccio replied:
Nah, this is ok. But a comment or something would help.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+		HTREEITEM _hItem;
+		bool _watch;
+
+		ProjectPanelData(DirectoryWatcher& directoryWatcher, const TCHAR* name, const TCHAR* filePath, NodeType nodeType, const std::vector<generic_string> filters = std::vector<generic_string>()) 
```
The `const TCHAR*` arguments should be one of:

- `_In_z_ const TCHAR*`
- `PCTSTR`

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+				_filePath = filePath;
+		}
+
+		virtual ~ProjectPanelData() {
```
Is this supposed to be an `override`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+		bool watchDir(bool watch) {
+			if (_hItem && _watch && !watch)
+				if (_nodeType == nodeType_baseDir || _nodeType == nodeType_dir)
+					_directoryWatcher.removeDir(_hItem);
```
How many levels of indentation are we at here? Braces please?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+		std::map<generic_string,HTREEITEM> _fileMap;
+	public:
+		TreeUpdaterDirectory( ProjectPanel *projectPanel, HTREEITEM hItem );
+		virtual ~TreeUpdaterDirectory() {}
```
`= default`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -103,46 +244,85 @@ class ProjectPanel : public DockingDlgInterface {
 		TreeView_SetTextColor(_treeView.getHSelf(), fgColour);
     };
 
+	TreeView& getTreeView() { return _treeView; }
```
Might as well also have a `const`-reference returning function?

```
const TreeView& getTreeView() const { return _treeView; }
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -167,4 +347,35 @@ private :
+	}
+
+
+    virtual void destroy() {
```
`= default`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -167,4 +347,35 @@ private :
+
+
+protected :
+	virtual INT_PTR CALLBACK run_dlgProc(UINT message, WPARAM wParam, LPARAM lParam);
```
Is this supposed to be an `override`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -102,15 +85,32 @@ HTREEITEM TreeView::addItem(const TCHAR *itemName, HTREEITEM hParentItem, int iI
+		tvItem.mask = TVIF_PARAM;
+		tvItem.hItem = hParentItem;
+		SendMessage(_hSelf, TVM_GETITEM, 0,(LPARAM)&tvItem);
+		TreeViewData* parentData = (TreeViewData*) tvItem.lParam;
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -102,15 +85,32 @@ HTREEITEM TreeView::addItem(const TCHAR *itemName, HTREEITEM hParentItem, int iI
+	}
+
+	HTREEITEM newItem = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	_validHandles.insert(newItem);
```
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -118,13 +118,20 @@ void TreeView::removeItem(HTREEITEM hTreeItem)
-	if (tvItem.lParam)
-		delete (generic_string *)(tvItem.lParam);
+	if (_listener)
+		_listener->onTreeItemRemoved(hTreeItem,(TreeViewData*) tvItem.lParam);
```
Oops, you missed a C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -118,13 +118,20 @@ void TreeView::removeItem(HTREEITEM hTreeItem)
 	TVITEM tvItem;
 	tvItem.hItem = hTreeItem;
 	tvItem.mask = TVIF_PARAM;
 	SendMessage(_hSelf, TVM_GETITEM, 0,(LPARAM)&tvItem);
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -118,13 +118,20 @@ void TreeView::removeItem(HTREEITEM hTreeItem)
+	// remove from valid handles
+	_validHandles.erase(hTreeItem);
+
+	delete (TreeViewData *)(tvItem.lParam);
```
Oops, you missed a C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -153,18 +172,22 @@ void TreeView::dupTree(HTREEITEM hTree2Dup, HTREEITEM hParentItem)
-			tvItem.lParam = (LPARAM)(new generic_string(*((generic_string *)(tvItem.lParam))));
-		}
+
+		tvItem.lParam = (LPARAM) ((TreeViewData*)tvItem.lParam)->clone();
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio replied:
You've also missed two C-style casts.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -197,14 +220,15 @@ void TreeView::cleanSubEntries(HTREEITEM hTreeItem)
+		if (_listener)
+			_listener->onTreeItemRemoved(hItem,(TreeViewData*)(tvItem.lParam));
+		_validHandles.erase(hItem);
+		delete (TreeViewData*)(tvItem.lParam);
```
Oops, you missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -355,8 +379,8 @@ void TreeView::moveTreeViewItem(HTREEITEM draggedItem, HTREEITEM targetItem)
-	if (tvDraggingItem.lParam)
-		tvDraggingItem.lParam = (LPARAM)(new generic_string(*((generic_string *)(tvDraggingItem.lParam))));
+
+	tvDraggingItem.lParam = (LPARAM) ((TreeViewData*)tvDraggingItem.lParam)->clone();
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio replied:
You've also missed two C-style casts.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -364,6 +388,10 @@ void TreeView::moveTreeViewItem(HTREEITEM draggedItem, HTREEITEM targetItem)
 	tvInsertStruct.hParent = targetItem;
 
 	HTREEITEM hTreeParent = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	_validHandles.insert(hTreeParent);
```
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -364,6 +388,10 @@ void TreeView::moveTreeViewItem(HTREEITEM draggedItem, HTREEITEM targetItem)
 	HTREEITEM hTreeParent = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	_validHandles.insert(hTreeParent);
+	if (_listener)
+		_listener->onTreeItemAdded(true, hTreeParent, (TreeViewData*) tvDraggingItem.lParam);
```
You've missed a C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -416,10 +444,8 @@ bool TreeView::swapTreeViewItem(HTREEITEM itemGoDown, HTREEITEM itemGoUp)
-	if (tvDownItem.lParam)
-		tvDownItem.lParam = (LPARAM)(new generic_string(*((generic_string *)(tvDownItem.lParam))));
+	tvUpItem.lParam = (LPARAM) ((TreeViewData*)tvUpItem.lParam)->clone();
+	tvDownItem.lParam = (LPARAM) ((TreeViewData*)tvDownItem.lParam)->clone();
```
You've missed a few more C-style casts.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -427,6 +453,9 @@ bool TreeView::swapTreeViewItem(HTREEITEM itemGoDown, HTREEITEM itemGoUp)
 	tvInsertUp.hInsertAfter = itemTop;
 	tvInsertUp.hParent = parentGoUp;
 	HTREEITEM hTreeParent1stInserted = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertUp);
+	_validHandles.insert(hTreeParent1stInserted);
```
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -427,6 +453,9 @@ bool TreeView::swapTreeViewItem(HTREEITEM itemGoDown, HTREEITEM itemGoUp)
 	HTREEITEM hTreeParent1stInserted = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertUp);
+	_validHandles.insert(hTreeParent1stInserted);
+	if (_listener)
+		_listener->onTreeItemAdded(true, hTreeParent1stInserted, (TreeViewData*) tvUpItem.lParam);
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -434,6 +463,9 @@ bool TreeView::swapTreeViewItem(HTREEITEM itemGoDown, HTREEITEM itemGoUp)
 	tvInsertDown.hInsertAfter = hTreeParent1stInserted;
 	tvInsertDown.hParent = parentGoDown;
 	HTREEITEM hTreeParent2ndInserted = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertDown);
+	_validHandles.insert(hTreeParent2ndInserted);
```
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -434,6 +463,9 @@ bool TreeView::swapTreeViewItem(HTREEITEM itemGoDown, HTREEITEM itemGoUp)
 	HTREEITEM hTreeParent2ndInserted = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertDown);
+	_validHandles.insert(hTreeParent2ndInserted);
+	if (_listener)
+		_listener->onTreeItemAdded(true, hTreeParent2ndInserted, (TreeViewData*) tvDownItem.lParam);
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -518,15 +550,16 @@ bool TreeView::searchLeafRecusivelyAndBuildTree(HTREEITEM tree2Build, const gene
-			{
-				tvItem.lParam = (LPARAM)(new generic_string(*((generic_string *)(tvItem.lParam))));
-			}
+			tvItem.lParam = (LPARAM) ((TreeViewData*) tvItem.lParam)->clone();
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -518,15 +550,16 @@ bool TreeView::searchLeafRecusivelyAndBuildTree(HTREEITEM tree2Build, const gene
 			tvInsertStruct.hInsertAfter = (HTREEITEM)TVI_LAST;
 			tvInsertStruct.hParent = tree2Build;
-			::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+			HTREEITEM hInsertedItem = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
```
You ***absolutely*** should check to see if `::SendMessage` returned `NULL` before using `newItem`.

***Especially*** since you're putting it in a variable named `_validHandles`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -518,15 +550,16 @@ bool TreeView::searchLeafRecusivelyAndBuildTree(HTREEITEM tree2Build, const gene
+			HTREEITEM hInsertedItem = (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+			_validHandles.insert(hInsertedItem);
+			if (_listener)
+				_listener->onTreeItemAdded(true, hInsertedItem, (TreeViewData*) tvItem.lParam);
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -560,10 +593,7 @@ bool TreeView::retrieveFoldingStateTo(TreeStateNode & treeState2Construct, HTREE
-	{
-		treeState2Construct._extraData = *((generic_string *)tvItem.lParam);
-	}
+	treeState2Construct._extraData = ((TreeViewData *)tvItem.lParam)->getExtraDataString();
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -591,11 +621,8 @@ bool TreeView::restoreFoldingStateFrom(const TreeStateNode & treeState2Compare,
-		if (treeState2Compare._extraData != *((generic_string *)tvItem.lParam))
-			return false;
-	}
+	if (treeState2Compare._extraData != ((TreeViewData *)tvItem.lParam)->getExtraDataString())
```
...and you've missed another C-style cast.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -619,10 +646,38 @@ bool TreeView::restoreFoldingStateFrom(const TreeStateNode & treeState2Compare,
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	SendMessage(_hSelf, TVM_GETITEM, 0,(LPARAM)&tvItem);
+	return (TreeViewData*) tvItem.lParam;
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -102,15 +85,32 @@ HTREEITEM TreeView::addItem(const TCHAR *itemName, HTREEITEM hParentItem, int iI
 	tvInsertStruct.hParent = hParentItem;
 
-	return (HTREEITEM)::SendMessage(_hSelf, TVM_INSERTITEM, 0, (LPARAM)(LPTVINSERTSTRUCT)&tvInsertStruct);
+	if (insertFunc)
```
Where does a caller actually pass an `insertFunc`?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.cpp:
```diff
@@ -197,14 +220,15 @@ void TreeView::cleanSubEntries(HTREEITEM hTreeItem)
-			delete (generic_string *)(tvItem.lParam);
-		}
+		if (_listener)
+			_listener->onTreeItemRemoved(hItem,(TreeViewData*)(tvItem.lParam));
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/TreeView.h:
```diff
@@ -88,6 +112,29 @@ class TreeView : public Window {
+		tvItem.hItem = hItem;
+		::SendMessage(_hSelf, TVM_GETITEM, 0, (LPARAM)&tvItem);
+
+		return (tvItem.state & TVIS_EXPANDED) != 0;
```
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.state`.

ariccio commented:
Sorry for the repetitive line notes! I saw several issues repeated, so I just copied and pasted the messages.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.cpp:
```diff
@@ -0,0 +1,389 @@
+		if (!_hStopEvent || !_hRunningEvent || !_hUpdateEvent)
+			throw std::runtime_error("Could not create DirectoryWatcher events");
+
+		_hThread = ::CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)threadFunc, (LPVOID)this, 0, NULL);
```
...Why are you casting to `LPVOID`? The standard guarantees that any pointer will implicitly convert to an equally `const`/`volatile` qualified `void*`. A need to do so explicitly means that there's a bug.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.cpp:
```diff
@@ -0,0 +1,389 @@
+
+		::WaitForSingleObject(_hRunningEvent, INFINITE);
+	}
+	catch (...)
```
Why the `catch (...)`?? You're only ever `throw`ing a `std::runtime_error`. Even worse, given `Notepad++`'s compilation options, this will also catch Access Violations!

Arkarit replied:
Sorry, I don't completely understand... I clean up and do a rethrow at the end of the catch. Why is this not ok (all other exceptions, also those not from me, will be rethrown)?

ariccio replied:
For reasons that are beyond me, `Notepad++` is compiled with [`/EHa`](https://msdn.microsoft.com/en-us/library/1deeycx5.aspx), which means that `catch(...)` does all sorts of surprising things - the docs say:

> Specifying /EHa and attempting to handle all exceptions by using catch(...) can be dangerous. In most cases, asynchronous exceptions are unrecoverable and should be considered fatal. Catching them and proceeding can cause process corruption and lead to bugs that are hard to find and fix.

> Why is this not ok (all other exceptions, also those not from me, will be rethrown)?

Before anything, I'm assuming that `Yuni::MutexLocker` doesn't throw. If it does, well...let me know.

You're only calling `WINAPI` functions here, none of which can throw.

If Windows *raises* an `SEH` exception, then program state is corrupted, and cleanup is nonsensical.

A `catch( std::runtime_error& )` deals with sane errors.

milipili replied:
The yuni library never throw. If it does, it is a bug.

damien-gerard replied:
For `/EHa`:
    C:\projects\npp\mili-npp\PowerEditor\src\MISC\Exception\Win32Exception.cpp(56): warning C4535: calling _set_se_translator() requires /EHa

ariccio replied:
Huh? Are you saying that `/EHa` isn't enabled? 

I just created a new directory, did a clean `git clone`, and checked the project settings under "C++" (or whatever that section of the property page is called) -> `Code Generation`, and `/EHa` was there?

milipili replied:
oh sorry. this warning comes from the cmake build, not from the current VS project.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+
+void ProjectPanel::onMessage(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+	wParam;
```
What the hell?

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -832,4 +842,42 @@ bool str2Clipboard(const generic_string &str2cpy, HWND hwnd)
+generic_string & stringTrimLeftInline(generic_string & result)
+{
+	result.erase(result.begin(), std::find_if(result.begin(), result.end(), std::not1(std::ptr_fun<int, int>(std::isspace))));
+	return result;
```
What's the point of *returning* the reference? I can't imagine that this is actually good for anything other than making alias analysis harder for the compiler.

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -832,4 +842,42 @@ bool str2Clipboard(const generic_string &str2cpy, HWND hwnd)
+generic_string stringTrim(const generic_string &s)
+{
+	generic_string result(s);
+	return stringTrimLeftInline(stringTrimRightInline(result));
```
Here, the `StringTrimLeftInline` returned reference is being copy constructed to a value..... why not just:


```
	generic_string result(s);
	stringTrimRightInline(result)
	stringTrimLeftInline(result);
	return result;
```

ariccio commented on PowerEditor/src/MISC/Common/Common.cpp:
```diff
@@ -832,4 +842,42 @@ bool str2Clipboard(const generic_string &str2cpy, HWND hwnd)
+
+generic_string & stringTrimInline(generic_string &s)
+{
+	return stringTrimLeftInline(stringTrimRightInline(s));
```
Same here.

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/Directory.h:
```diff
@@ -0,0 +1,126 @@
+
+	struct comparator {
+		bool operator() (const generic_string& lhs, const generic_string& rhs) const {
+			return lstrcmpi(lhs.c_str(),rhs.c_str()) < 0;
```
`_tcsicmp` please?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.h:
```diff
@@ -0,0 +1,154 @@
+	void removeDir(HTREEITEM treeItem);
+	void removeAllDirs();
+
+	bool getManualMode() const { 
```
Where is `getManualMode` called from?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.h:
```diff
@@ -0,0 +1,154 @@
+	}
+	void setManualMode(bool val);
+
+	HWND getWindow() const { 
```
Where is `getWindow` called from?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
 					return false;
 			}
 		}
+		if (lstrcmp(TEXT("Directory"), v) == 0)
```
`_tcscmp` please!

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp:
```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
 	return true;
 }
 
+void ProjectPanel::rebuildDirectoryTree(HTREEITEM hParentItem, const ProjectPanelData& data)
```
[Here's a neat trick](http://ariccio.com/2015/04/02/preventing-bugs-and-improving-code-quality-with-microsoft-sal-part-2-custom-preconditions-for-structs-objects/) to help detect (via static analysis) conditions that might lead to the `assert(0);` at line 692:

```
_Pre_satisfies_( ( data._nodeType == nodeType_baseDir ) || ( data._nodeType == nodeType_dir ) )
void ProjectPanel::rebuildDirectoryTree(HTREEITEM hParentItem, const ProjectPanelData& data)
```

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+	{
+	};
+
+	virtual ~ProjectPanel() {
```
Supposed to be an `= 0;` function?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.h:
```diff
@@ -63,24 +69,159 @@
+
+	friend int CALLBACK compareFunc(LPARAM, LPARAM, LPARAM);
+
+	enum NodeType { nodeType_dummy = -1, nodeType_root = 0, nodeType_project = 1, nodeType_folder = 2, nodeType_file = 3, nodeType_baseDir = 4, nodeType_dir = 5, nodeType_dirFile = 6, };
```
maybe an `enum class NodeType` would be better?

ariccio commented on PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.cpp:
```diff
@@ -0,0 +1,389 @@
+void DirectoryWatcher::iterateDirs()
+{
+
+	std::map<Directory*,bool> changedMap;
```
Does this map need to be ordered? `std::unordered_map` can be *a __lot faster__*.

ariccio commented:
Some general thoughts:
- Whenever you're calling across an API boundary, into a function that has *out* parameters, and **not** properly checking the return code, you're triggering undefined behavior. Unless it's otherwise explicitly specified, then the state of the *out* parameters is undefined. Ignoring return values can often *work when tested*, but later, somewhere down the road, causes sinister bugs.
- I didn't see that many `assert`s. They're not strictly necessary, and you're free to ignore this, but it's always a good idea to `assert` like you're being paid-per-`assert`.

Some thoughts specifically about this Pull Request:

I like the idea, but in its current form, it probably won't be accepted - it's too big & too complex. 

- It seems like there's a lot of duplicated code. Try refactoring some of it?
- Many control structures are just ***far too complex***. I'm getting a headache while just doing a cursory review of them, and that doesn't bode well for merging.
- Move some code out of the headers? It might just be me, but I think it's harder to read headers that have lots of inline code. I consider `virtual` functions an exception to this, because defining them in headers can greatly aid in devirtualization. *Personally*, I question the value of code like:

```
bool getManualMode() const { 
		return _manualMode; 
	}
```
- Look into [`ReadDirectoryChanges`](http://qualapps.blogspot.com/2010/05/understanding-readdirectorychangesw.html). I share your hatred of polling.
- Try to cut the number of raw changes in half. Consider it a challenge, if you're having trouble.
- Refactor code that looks like (the following) into a separate function. An `anonymous` namespace is a perfect place for a free function containing the factored-out code. i.e.:


*imagine the conditions are more complex*
```
//some large function above
int someVariable;

if ( someCondition )
    someVariable = 1;
else
    someVariable = 0;
//rest of some large function
```
...to:
```
namespace {

int someDescriptiveName( bool someCondition )
{
    if ( someCondition )
        return 1;
    else
        return 0;
}

}
```

...so that the caller can just:

```
const int someVariable = someDescriptiveName( someCondition );
```

Arkarit commented:
Thank you very much for your review, ariccio!
And sorry for the huge amount of problems and for the late answer (got a lot of work to do, and I really underestimated this directory thing...)

A lot of problems were caused by C&P from old code. As far as I can see e.g. the SendMessage() without checks is also done in the master branch all over the place in TreeView and ProjectPanel.
Also, there's also a lot of duplicated code in the master as well.
I think it's the best way to rework these problems completely and consistently and place all of this duplicated code into some functions with checking.

Regarding ReadDirectoryChanges I agree. It would be definitely better to use this, even if the directory will be locked (but not its sub directories). Nevertheless, I will first fix the problems mentioned here and then look into this.

milipili commented:
From my point of view, you should wait before spending too much time on this. As soon as it compiles with cmake there will have several incompatible changes and this code in the current state won't be isolated enough.

Arkarit commented:
OK, thanks for the hint. I have already begun with the changes, but I will wait until that then.
Is there an estimation when this will be done?

milipili commented:
The compilation with cmake is on-going and should be here soon. For the rest I don't know.

ariccio commented:
> And sorry for the huge amount of problems and for the late answer (got a lot of work to do, and I really underestimated this directory thing...)

Lol, no worries!

> A lot of problems were caused by C&P from old code. As far as I can see e.g. the SendMessage() without checks is also done in the master branch all over the place in TreeView and ProjectPanel.
> Also, there's also a lot of duplicated code in the master as well.
> I think it's the best way to rework these problems completely and consistently and place all of this duplicated code into some functions with checking.

Agreed, but we need to start somewhere.



> From my point of view, you should wait before spending too much time on this. As soon as it compiles with cmake there will have several incompatible changes and this code in the current state won't be isolated enough.

*mostly* agreed ;)

We can still ponder changes in the meantime. Any ideas what changes will be incompatible?

milipili commented:
all files will be progressively rewritten then moved to a new folder.

donho commented:
Without mentioning my personal dislike of polling per second, any PR should be implemented with a accepted feature request / bug repport. Otherwise your PR will take a great risk to be rejected.
